contrib/masterfiles-stage/common.sh

```diff
@@ -12,175 +12,252 @@ if [ ! "$debug_mode" == "true" ]; then
+                                 "${1}/cf_promises_release_id" ; then
+      # release id is the same in stage and deploy dir
+      # so prevent triggering update on hosts by keeping old "validated" flag file
+      cp -a "${1}/cf_promises_validated" "${temp_stage}/"
+    fi
+    local third_dir
+    third_dir="$(mktemp -d --tmpdir="$(dirname "$1")" )"
+    mv "${1}" "${third_dir}"  || error_exit "Can't mv ${1} to ${third_dir}"
+    mv "${temp_stage}" "${1}"          || error_exit "Can't mv ${temp_stage} to ${1}"
+    rm -rf "${third_dir}"
   fi
+
+  # Note about triggering policy updates:
+  #
+  # cf_promises_validated gets updated by any run of cf-promises,
+  # but hosts use cf_promises_validated as the flag file to see
+  # if they need to update everything else (the full policy set.)
+  #
+  # cf_promises_release_id is the same for a given policy set
+  # unless changes have actually been made to the policy, so it
+  # can be used to check if we want to trigger an update.
+  #
+  # In other words, update is triggered by putting the
+  # newly created copy of cf_promises_validated into the MASTERDIR
+  # and update is avoided either by:
+  #
+  # 1. Completely skipping the rollout_staged_policy_to_masterdir
+  # function, or
+  #
+  # 2. Copying the MASTERDIR's copy of cf_promises_validated
+  # *back* into the STAGING_DIR *before* performing the rollout,
+  # so that after the rollout the MASTERDIR's copy of the flag
+  # file is the same as it was before the rollout.
+  #
+  # This function uses the second approach.  --Mike Weilgart
+}
+
+######################################################
+##           VCS_TYPE-based main functions           #
+######################################################
+
+git_stage_policy_channels() {
+# Contributed by Mike Weilgart
+
+  # Depends on ${channel_config[@]} and $dir_to_hold_mirror
+  # Calls functions dependent on $GIT_URL
+  # (See the example git policy channels params file.)
+  #
+  # Stages multiple policy channels from a specified GIT_URL,
+  # each to the specified path.
+  #
+  # The paths to stage to as well as the policy sets to stage
+  # are both specified in the "channel_config" array in the
+  # PARAMS file.
+  #
+  # The value of MASTERDIR that is assigned in masterfiles-stage.sh
+  # is ignored by this function, since there is effectively a separate
+  # MASTERDIR for each separate policy channel.
+  #
+  # Example value for channel_config:
+  # (This is a single value, split into three lines for readability.)
+  #
+  # channel_config=()
+  # channel_config+=( "/var/cfengine/masterfiles" "my_git_tag" )
+  # channel_config+=( "/var/cfengine/policy_channels/channel_1" "my_git_branch" )
+
+  # Simplest validation check first
+  set -- "${channel_config[@]}"
+  [ "$#" -gt 1 ] ||
+    error_exit "The channel_config array must have at least two elements."
+
+  check_git_installed
+  git_setup_local_mirrored_repo "$dir_to_hold_mirror"
```

Comment:
Note the new parameter `$dir_to_hold_mirror` to be set in the PARAMS file.  This is so the policy channels framework doesn't use ROOT in any way.